File: apps/api/app/schemas/admin_feature_request.py
"""Admin feature-request 검토 큐 schema — `docs/api/admin.md` (T-179).

사용자 feature 제안(`app.feature_suggestions`)을 Admin이 검토·승인/거절하는 화면용.
승인 시 kor_travel_map `/v1/admin/features*` change API로 전송(전송 client = T-180).
"""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field

from app.schemas.feature import (
    Coord,
    ExternalRef,
    FeatureKind,
    FeatureRequestStatus,
    FeatureRequestType,
)


class AdminFeatureRequestSummary(BaseModel):
    """검토 큐 1건 (사용자 이메일은 마스킹)."""

    request_id: uuid.UUID
    requester_user_id: uuid.UUID
    requester_email_masked: str | None = None
    type: FeatureRequestType
    kind: FeatureKind
    name: str
    coord: Coord
    categories: list[str] = Field(default_factory=list)
    note: str | None = None
    target_feature_id: str | None = None
    source: str = "user"
    external_ref: ExternalRef | None = None
    status: FeatureRequestStatus
    kor_travel_map_ref: dict[str, Any] | None = None
    reviewed_by_admin_id: uuid.UUID | None = None
    created_at: datetime
    resolved_at: datetime | None = None


class AdminFeatureRequestPagedResponse(BaseModel):
    items: list[AdminFeatureRequestSummary] = Field(default_factory=list)
    total: int
    page: int
    limit: int


class AdminFeatureRequestApprove(BaseModel):
    """승인 입력.

    new_place는 범용 Map 요청 큐에 suggestion UUID 그대로 제출한다. ``category``가 있으면 기존
    suggestion category 목록에 중복 없이 보강한다. ``marker_*``는 correction override 호환 입력이며
    Map 큐의 최종 승인자가 정한다. ``access_reason``은 PinVi audit 사유, ``kor_travel_map_reason``은
    correction/closure change request 사유(미지정 시 access_reason)다.
    """

    access_reason: str = Field(min_length=1, max_length=500)
    category: str | None = Field(default=None, max_length=32)
    marker_color: str | None = Field(default=None, pattern=r"^P-\d{2}$")
    marker_icon: str | None = Field(default=None, max_length=64)
    name: str | None = Field(default=None, min_length=1, max_length=200)
    kor_travel_map_reason: str | None = Field(default=None, max_length=500)


class AdminFeatureRequestReject(BaseModel):
    access_reason: str = Field(min_length=1, max_length=500)


class AdminFeatureRequestResult(BaseModel):
    """승인/거절 후 갱신된 상태."""

    request_id: uuid.UUID
    status: FeatureRequestStatus
    kor_travel_map_ref: dict[str, Any] | None = None
    reviewed_by_admin_id: uuid.UUID | None = None
    resolved_at: datetime | None = None

File: packages/schemas/src/admin_feature_request.ts
import { z } from 'zod';
import { CoordSchema, Iso8601Schema } from './common';
import {
  ExternalRefSchema,
  FeatureKindSchema,
  FeatureRequestStatusSchema,
  FeatureRequestTypeSchema,
} from './feature';

/** Admin feature-request 검토 큐 1건 (요청자 이메일은 마스킹). */
export const AdminFeatureRequestSummarySchema = z.object({
  request_id: z.string().uuid(),
  requester_user_id: z.string().uuid(),
  requester_email_masked: z.string().nullable().optional(),
  type: FeatureRequestTypeSchema,
  kind: FeatureKindSchema,
  name: z.string(),
  coord: CoordSchema,
  categories: z.array(z.string()),
  note: z.string().nullable().optional(),
  target_feature_id: z.string().nullable().optional(),
  source: z.string().default('user'),
  external_ref: ExternalRefSchema.nullable().default(null),
  status: FeatureRequestStatusSchema,
  kor_travel_map_ref: z.record(z.string(), z.unknown()).nullable().optional(),
  reviewed_by_admin_id: z.string().uuid().nullable().optional(),
  created_at: Iso8601Schema,
  resolved_at: Iso8601Schema.nullable().optional(),
});
export type AdminFeatureRequestSummary = z.infer<typeof AdminFeatureRequestSummarySchema>;

export const AdminFeatureRequestPagedResponseSchema = z.object({
  items: z.array(AdminFeatureRequestSummarySchema),
  total: z.number().int(),
  page: z.number().int(),
  limit: z.number().int(),
});
export type AdminFeatureRequestPagedResponse = z.infer<
  typeof AdminFeatureRequestPagedResponseSchema
>;

/** 승인 입력 — new_place는 category(8자리)/marker_color/marker_icon이 필수(서버 422). */
export const AdminFeatureRequestApproveSchema = z.object({
  access_reason: z.string().min(1).max(500),
  category: z.string().max(32).optional(),
  marker_color: z
    .string()
    .regex(/^P-\d{2}$/, 'marker color는 P-01~P-16 형식.')
    .optional(),
  marker_icon: z.string().max(64).optional(),
  name: z.string().min(1).max(200).optional(),
  kor_travel_map_reason: z.string().max(500).optional(),
});
export type AdminFeatureRequestApprove = z.infer<typeof AdminFeatureRequestApproveSchema>;

export const AdminFeatureRequestRejectSchema = z.object({
  access_reason: z.string().min(1).max(500),
});
export type AdminFeatureRequestReject = z.infer<typeof AdminFeatureRequestRejectSchema>;

export const AdminFeatureRequestResultSchema = z.object({
  request_id: z.string().uuid(),
  status: FeatureRequestStatusSchema,
  kor_travel_map_ref: z.record(z.string(), z.unknown()).nullable().optional(),
  reviewed_by_admin_id: z.string().uuid().nullable().optional(),
  resolved_at: Iso8601Schema.nullable().optional(),
});
export type AdminFeatureRequestResult = z.infer<typeof AdminFeatureRequestResultSchema>;
